File: Src/java/cql-to-elm/src/test/java/org/cqframework/cql/cql2elm/LiteralTests.java
package org.cqframework.cql.cql2elm;

import org.hl7.elm.r1.*;
import org.testng.annotations.Test;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

import static org.cqframework.cql.cql2elm.matchers.HasTypeAndResult.hasTypeAndResult;
import static org.cqframework.cql.cql2elm.matchers.ListOfLiterals.listOfLiterals;
import static org.cqframework.cql.cql2elm.matchers.LiteralFor.literalFor;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.Assert.assertEquals;

/**
 * Created by Bryn on 11/21/2017.
 */
public class LiteralTests {

    private Map<String, ExpressionDef> defs;

    @Test
    public void dateTimeLiteralTests() throws IOException {
        CqlTranslator translator = TestUtils.runSemanticTest("DateTimeLiteralTest.cql", 0);
        Library library = translator.toELM();
        defs = new HashMap<>();
        if (library.getStatements() != null) {
            for (ExpressionDef def : library.getStatements().getDef()) {
                defs.put(def.getName(), def);
            }
        }

        ExpressionDef def = defs.get("TimeZoneDateTimeLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));
        DateTime dateTime = (DateTime)def.getExpression();
        assertThat(dateTime.getTimezoneOffset(), literalFor(-7.0));

        def = defs.get("TimeZonePositiveDateTimeLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));
        dateTime = (DateTime)def.getExpression();
        assertThat(dateTime.getTimezoneOffset(), literalFor(7.0));

        def = defs.get("YearLiteral");
        assertThat(def, hasTypeAndResult(Date.class, "System.Date"));

        def = defs.get("DateTimeYearLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));

        def = defs.get("UTCYearLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));

        def = defs.get("YearMonthLiteral");
        assertThat(def, hasTypeAndResult(Date.class, "System.Date"));

        def = defs.get("DateTimeYearMonthLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));

        def = defs.get("UTCYearMonthLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));

        def = defs.get("DateLiteral");
        assertThat(def, hasTypeAndResult(Date.class, "System.Date"));

        def = defs.get("DateTimeDateLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));

        def = defs.get("UTCDateLiteral");
        assertThat(def, hasTypeAndResult(DateTime.class, "System.DateTime"));
    }

    @Test
    public void quantityLiteralTests() throws IOException {
        CqlTranslator translator = TestUtils.runSemanticTest("QuantityLiteralTest.cql", 1);
        Library library = translator.toELM();
        defs = new HashMap<>();
        if (library.getStatements() != null) {
            for (ExpressionDef def : library.getStatements().getDef()) {
                defs.put(def.getName(), def);
            }
        }

        ExpressionDef def = defs.get("ValidQuantityLiteral");
        assertThat(def, hasTypeAndResult(Quantity.class, "System.Quantity"));
        Quantity quantity = (Quantity)def.getExpression();
        assertThat(quantity.getValue(), is(BigDecimal.valueOf(10)));
        assertThat(quantity.getUnit(), is("mm[Hg]"));

        def = defs.get("InvalidQuantityLiteral");
        assertThat("Invalid quantity literal is returned as a Null", def.getExpression() instanceof Null);

        def = defs.get("UnitQuantityLiteral");
        assertThat(def, hasTypeAndResult(Quantity.class, "System.Quantity"));
        quantity = (Quantity)def.getExpression();
        assertThat(quantity.getValue(), is(BigDecimal.valueOf(10)));
        assertThat(quantity.getUnit(), is("1"));

        def = defs.get("AnnotationQuantityLiteral");
        assertThat(def, hasTypeAndResult(Quantity.class, "System.Quantity"));
        quantity = (Quantity)def.getExpression();
        assertThat(quantity.getValue(), is(BigDecimal.valueOf(10)));
        assertThat(quantity.getUnit(), is("{shab-shab-shab}"));

        def = defs.get("QuantityConversionTest");
        assertThat(def, hasTypeAndResult(ConvertQuantity.class, "System.Quantity"));
        ConvertQuantity convertQuantity = (ConvertQuantity)def.getExpression();
        assertThat(convertQuantity.getOperand().get(0), instanceOf(Quantity.class));
        quantity = (Quantity)convertQuantity.getOperand().get(0);
        assertThat(quantity.getValue(), is(BigDecimal.valueOf(5)));
        assertThat(quantity.getUnit(), is("mg"));
        assertThat(convertQuantity.getOperand().get(1), instanceOf(Literal.class));
        Literal literal = (Literal)convertQuantity.getOperand().get(1);
        assertThat(literal.getValue(), is("g"));

        def = defs.get("QuantityConversionWeekTest");
        assertThat(def, hasTypeAndResult(ConvertQuantity.class, "System.Quantity"));
        convertQuantity = (ConvertQuantity)def.getExpression();
        assertThat(convertQuantity.getOperand().get(0), instanceOf(Quantity.class));
        quantity = (Quantity)convertQuantity.getOperand().get(0);
        assertThat(quantity.getValue(), is(BigDecimal.valueOf(28)));
        assertThat(quantity.getUnit(), is("days"));
        assertThat(convertQuantity.getOperand().get(1), instanceOf(Literal.class));
        literal = (Literal)convertQuantity.getOperand().get(1);
        assertThat(literal.getValue(), is("wk"));
    }

    private Map<String, ExpressionDef> getDefs(Library library) {
        Map<String, ExpressionDef> result = new HashMap<>();
        if (library.getStatements() != null) {
            for (ExpressionDef def : library.getStatements().getDef()) {
                result.put(def.getName(), def);
            }
        }
        return result;
    }

    @Test
    public void RatioLiteralTests() throws IOException {
        CqlTranslator translator = TestUtils.runSemanticTest("RatioLiteralTest.cql", 0);
        Library library = translator.toELM();
        defs = getDefs(library);

        ExpressionDef def = defs.get("SimpleRatio");
        assertThat(def, hasTypeAndResult(Ratio.class, "System.Ratio"));
        Ratio ratio = (Ratio)def.getExpression();
        assertThat(ratio.getNumerator().getValue(), is(BigDecimal.valueOf(5)));
        assertThat(ratio.getDenominator().getValue(), is(BigDecimal.valueOf(5)));

        def = defs.get("QuantityRatio");
        assertThat(def, hasTypeAndResult(Ratio.class, "System.Ratio"));
        ratio = (Ratio)def.getExpression();
        assertThat(ratio.getNumerator().getValue(), is(BigDecimal.valueOf(5)));
        assertThat(ratio.getNumerator().getUnit(), is("mg"));
        assertThat(ratio.getDenominator().getValue(), is(BigDecimal.valueOf(100)));
        assertThat(ratio.getDenominator().getUnit(), is("mL"));
    }

    @Test
    public void testDecimal() throws IOException {
        CqlTranslator translator = TestUtils.createTranslatorFromText("define TestDecimal: 1.5");
        Library library = translator.toELM();
        defs = getDefs(library);

        ExpressionDef def = defs.get("TestDecimal");
        assertThat(def, hasTypeAndResult(Literal.class, "System.Decimal"));

        Literal literal = (Literal)def.getExpression();
        assertEquals("1.5", literal.getValue());
    }

    @Test
    public void testString() throws IOException {
        CqlTranslator translator = TestUtils.createTranslatorFromText("define TestString: '12345''");
        Library library = translator.toELM();
        defs = getDefs(library);

        ExpressionDef def = defs.get("TestString");
        assertThat(def, hasTypeAndResult(Literal.class, "System.String"));

        Literal literal = (Literal)def.getExpression();
        assertEquals("12345", literal.getValue());
    }

    @Test
    public void testInteger() throws IOException {
        CqlTranslator translator = TestUtils.createTranslatorFromText("define TestInteger: 12345");
        Library library = translator.toELM();
        defs = getDefs(library);

        ExpressionDef def = defs.get("TestInteger");
        assertThat(def, hasTypeAndResult(Literal.class, "System.Integer"));

        Literal literal = (Literal)def.getExpression();
        assertEquals("12345", literal.getValue());
    }

    @Test
    public void testLongInteger() throws IOException {
        CqlTranslator translator = TestUtils.createTranslatorFromText("define TestLongInteger: 12345L");
        Library library = translator.toELM();
        defs = getDefs(library);

        ExpressionDef def = defs.get("TestLongInteger");
        assertThat(def, hasTypeAndResult(Literal.class, "System.Long"));

        Literal literal = (Literal)def.getExpression();
        assertEquals("12345", literal.getValue());

    }

    @Test
    public void TokenRecognitionErrorTest() throws IOException {
        CqlTranslator translator = TestUtils.runSemanticTest("TokenRecognitionError.cql", 1);
        assertThat(translator.getErrors().size(), equalTo(1));
    }
}

File: Src/java/cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/LiteralTests.kt
package org.cqframework.cql.cql2elm

import java.io.IOException
import java.math.BigDecimal
import kotlin.collections.set
import org.cqframework.cql.cql2elm.matchers.HasTypeAndResult
import org.cqframework.cql.cql2elm.matchers.LiteralFor.Companion.literalFor
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers
import org.hl7.elm.r1.ConvertQuantity
import org.hl7.elm.r1.Date
import org.hl7.elm.r1.DateTime
import org.hl7.elm.r1.Expression
import org.hl7.elm.r1.ExpressionDef
import org.hl7.elm.r1.Library
import org.hl7.elm.r1.Literal
import org.hl7.elm.r1.Null
import org.hl7.elm.r1.Quantity
import org.hl7.elm.r1.Ratio
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

@Suppress("LongMethod")
/** Created by Bryn on 11/21/2017. */
internal class LiteralTests {
    @Test
    @Throws(IOException::class)
    fun dateTimeLiteralTests() {
        val translator = TestUtils.runSemanticTest("DateTimeLiteralTest.cql", 0)
        val library = translator.toELM()
        defs = HashMap()
        if (library!!.statements != null) {
            for (def in library.statements!!.def) {
                defs!![def.name] = def
            }
        }

        var def: ExpressionDef = defs!!["TimeZoneDateTimeLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))
        var dateTime = def.expression as DateTime?
        assertThat<Expression?>(dateTime!!.timezoneOffset, literalFor(-7.0))

        def = defs!!["TimeZonePositiveDateTimeLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))
        dateTime = def.expression as DateTime?
        assertThat<Expression?>(dateTime!!.timezoneOffset, literalFor(7.0))

        def = defs!!["YearLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Date::class.java, "System.Date"))

        def = defs!!["DateTimeYearLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))

        def = defs!!["UTCYearLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))

        def = defs!!["YearMonthLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Date::class.java, "System.Date"))

        def = defs!!["DateTimeYearMonthLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))

        def = defs!!["UTCYearMonthLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))

        def = defs!!["DateLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Date::class.java, "System.Date"))

        def = defs!!["DateTimeDateLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))

        def = defs!!["UTCDateLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))

        def = defs!!["TimeZoneHalfHourLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(DateTime::class.java, "System.DateTime"))
        dateTime = def.expression as DateTime?
        assertThat<Expression?>(dateTime!!.timezoneOffset, literalFor(1.5))
    }

    @Test
    @Throws(IOException::class)
    fun quantityLiteralTests() {
        val translator = TestUtils.runSemanticTest("QuantityLiteralTest.cql", 1)
        val library = translator.toELM()
        defs = HashMap()
        if (library!!.statements != null) {
            for (def in library.statements!!.def) {
                defs!![def.name] = def
            }
        }

        var def: ExpressionDef = defs!!["ValidQuantityLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Quantity::class.java, "System.Quantity"))
        var quantity = def.expression as Quantity?
        assertThat<BigDecimal?>(
            quantity!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(10))
        )
        assertThat<String?>(quantity.unit, Matchers.`is`<String?>("mm[Hg]"))

        def = defs!!["InvalidQuantityLiteral"]!!
        assertThat("Invalid quantity literal is returned as a Null", def.expression is Null)

        def = defs!!["UnitQuantityLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Quantity::class.java, "System.Quantity"))
        quantity = def.expression as Quantity?
        assertThat<BigDecimal?>(
            quantity!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(10))
        )
        assertThat<String?>(quantity.unit, Matchers.`is`<String?>("1"))

        def = defs!!["AnnotationQuantityLiteral"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Quantity::class.java, "System.Quantity"))
        quantity = def.expression as Quantity?
        assertThat<BigDecimal?>(
            quantity!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(10))
        )
        assertThat<String?>(quantity.unit, Matchers.`is`<String?>("{shab-shab-shab}"))

        def = defs!!["QuantityConversionTest"]!!
        assertThat(
            def,
            HasTypeAndResult.hasTypeAndResult(ConvertQuantity::class.java, "System.Quantity")
        )
        var convertQuantity = def.expression as ConvertQuantity?
        assertThat(convertQuantity!!.operand[0], Matchers.instanceOf(Quantity::class.java))
        quantity = convertQuantity.operand[0] as Quantity?
        assertThat<BigDecimal?>(quantity!!.value, Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(5)))
        assertThat<String?>(quantity.unit, Matchers.`is`<String?>("mg"))
        assertThat(convertQuantity.operand[1], Matchers.instanceOf(Literal::class.java))
        var literal: Literal = convertQuantity.operand[1] as Literal
        assertThat(literal.value, Matchers.`is`("g"))

        def = defs!!["QuantityConversionWeekTest"]!!
        assertThat(
            def,
            HasTypeAndResult.hasTypeAndResult(ConvertQuantity::class.java, "System.Quantity")
        )
        convertQuantity = def.expression as ConvertQuantity?
        assertThat(convertQuantity!!.operand[0], Matchers.instanceOf(Quantity::class.java))
        quantity = convertQuantity.operand[0] as Quantity?
        assertThat<BigDecimal?>(
            quantity!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(28))
        )
        assertThat<String?>(quantity.unit, Matchers.`is`<String?>("days"))
        assertThat(convertQuantity.operand[1], Matchers.instanceOf(Literal::class.java))
        literal = convertQuantity.operand[1] as Literal
        assertThat(literal.value, Matchers.`is`("wk"))
    }

    private fun getDefs(library: Library): MutableMap<String?, ExpressionDef> {
        val result: MutableMap<String?, ExpressionDef> = HashMap()
        if (library.statements != null) {
            for (def in library.statements!!.def) {
                result[def.name] = def
            }
        }
        return result
    }

    @Test
    @Throws(IOException::class)
    fun ratioLiteralTests() {
        val translator = TestUtils.runSemanticTest("RatioLiteralTest.cql", 0)
        val library = translator.toELM()
        defs = getDefs(library!!)

        var def: ExpressionDef = defs!!["SimpleRatio"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Ratio::class.java, "System.Ratio"))
        var ratio = def.expression as Ratio?
        assertThat<BigDecimal?>(
            ratio!!.numerator!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(5))
        )
        assertThat<BigDecimal?>(
            ratio.denominator!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(5))
        )

        def = defs!!["QuantityRatio"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Ratio::class.java, "System.Ratio"))
        ratio = def.expression as Ratio?
        assertThat<BigDecimal?>(
            ratio!!.numerator!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(5))
        )
        assertThat<String?>(ratio.numerator!!.unit, Matchers.`is`<String?>("mg"))
        assertThat<BigDecimal?>(
            ratio.denominator!!.value,
            Matchers.`is`<BigDecimal?>(BigDecimal.valueOf(100))
        )
        assertThat<String?>(ratio.denominator!!.unit, Matchers.`is`<String?>("mL"))
    }

    @Test
    @Throws(IOException::class)
    fun decimal() {
        val translator = TestUtils.createTranslatorFromText("define TestDecimal: 1.5")
        val library = translator.toELM()
        defs = getDefs(library!!)

        val def: ExpressionDef = defs!!["TestDecimal"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Literal::class.java, "System.Decimal"))

        val literal = def.expression as Literal?
        Assertions.assertEquals("1.5", literal!!.value)
    }

    @Test
    @Throws(IOException::class)
    fun string() {
        val translator = TestUtils.createTranslatorFromText("define TestString: '12345''")
        val library = translator.toELM()
        defs = getDefs(library!!)

        val def: ExpressionDef = defs!!["TestString"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Literal::class.java, "System.String"))

        val literal = def.expression as Literal?
        Assertions.assertEquals("12345", literal!!.value)
    }

    @Test
    @Throws(IOException::class)
    fun integer() {
        val translator = TestUtils.createTranslatorFromText("define TestInteger: 12345")
        val library = translator.toELM()
        defs = getDefs(library!!)

        val def: ExpressionDef = defs!!["TestInteger"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Literal::class.java, "System.Integer"))

        val literal = def.expression as Literal?
        Assertions.assertEquals("12345", literal!!.value)
    }

    @Test
    @Throws(IOException::class)
    fun longInteger() {
        val translator = TestUtils.createTranslatorFromText("define TestLongInteger: 12345L")
        val library = translator.toELM()
        defs = getDefs(library!!)

        val def: ExpressionDef = defs!!["TestLongInteger"]!!
        assertThat(def, HasTypeAndResult.hasTypeAndResult(Literal::class.java, "System.Long"))

        val literal = def.expression as Literal?
        Assertions.assertEquals("12345", literal!!.value)
    }

    @Test
    @Throws(IOException::class)
    fun tokenRecognitionErrorTest() {
        val translator = TestUtils.runSemanticTest("TokenRecognitionError.cql", 1)
        assertThat(translator.errors.size, Matchers.equalTo(1))
    }

    companion object {
        private var defs: MutableMap<String?, ExpressionDef>? = null
    }
}
